File: 백준/Gold/11401. 이항 계수 3/이항 계수 3.py
P = 1_000_000_007

def factorial(n):
    ret = 1
    for i in range(n, 1, -1):
        ret = (ret * i) % P
    return ret
        
def pow(base, exp):
    ret = 1
    while exp > 0:
        if exp % 2 == 1:
            ret *= base
            ret %= P
        base = (base * base) % P
        exp //= 2
    return ret

N, K = map(int, input().split(" "))

denom = factorial(K) * factorial(N-K) % P
print(factorial(N) * pow(denom, P - 2) % P)

File: 백준/Gold/11401. 이항 계수 3/이항 계수 3.rs
// test performance

const P: i64 = 1_000_000_007;

fn main() {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).expect("Failed to read input");
    let mut parts = input.trim().split_whitespace().map(|s| s.parse::<i64>().expect("Invalid input"));
    
    let n = parts.next().unwrap();
    let k = parts.next().unwrap();
    
    let denom = (factorial(k) * factorial(n - k)) % P;
    println!("{}", (factorial(n) * pow(denom, P - 2)) % P);
}

fn factorial(n: i64) -> i64 {
    let mut result = 1;
    for i in 1..=n {
        result *= i;
        result %= P;
    }
    result
}

fn pow(base: i64, exp: i64) -> i64 {
    let mut ret = 1;
    let mut base = base;
    let mut exp = exp;
    
    while exp > 0 {
        if exp % 2 == 1 {
            ret *= base;
            ret %= P;
        }
        base = (base * base) % P;
        exp /= 2;
    }
    ret
}
